### PR TITLE
Db 703 flex fields

### PR DESCRIPTION
**Pull request by michaeldhess**
Allowing users to submit files with "_flex" columns that are added to the error reports.

- Read "flex_" columns into flex dictionary
- Store columns in flex_fields table
- Add flex data to error reports

### CODE REVIEW

**Comment by klrBAH** [comment]
Can one of the admins verify this patch?

**Comment by klrBAH** [comment]
[run tests]

File: dataactcore/models/stagingModels.py [hunk 1]
```diff
@@ -21,6 +21,22 @@ def concatTas(context):
...
+    job_id = Column(Integer, nullable=False, index=True)
+    row_number = Column(Integer, nullable=False)
+    header = Column(Text)
+    cell = Column(Text)
+    
+    def __init__(self, **kwargs):
+        # broker is set up to ignore extra columns in submitted data
+        # so get rid of any extraneous kwargs before instantiating
+        cleanKwargs = {k: v for k, v in kwargs.items() if hasattr(self, k)}
+        super(FlexField, self).__init__(**cleanKwargs)
```
Comment (cmc333333) [inline dataactcore/models/stagingModels.py]: Is this required? We have it on another model as a convenience, but hopefully we don't need it here as well.

File: dataactvalidator/validation_handlers/validationManager.py [hunk 2]
```diff
@@ -167,8 +174,9 @@ def readRecord(self,reader,writer,file_type,interfaces,row_number,job_id,fields,
                 writer.write(["Formatting Error", ValidationError.readErrorMsg, str(row_number), ""])
                 error_list.recordRowError(job_id,self.filename,"Formatting Error",ValidationError.readError,row_number,severity_id=interfaces.validationDb.getRuleSeverityId("fatal"))
                 row_error_found = True
-            return {}, reduce_row, True, False, row_error_found
-        return record, reduce_row, False, False, row_error_found
+
+            return {}, reduce_row, True, False, row_error_found, None
+        return record, reduce_row, False, False, row_error_found, flex_cols
```
Comment (cmc333333) [inline dataactvalidator/validation_handlers/validationManager.py]: Given how many args we're returning, would it make sense to define a `namedtuple` to make this a bit easier to read? Relatedly, would you mind modifying the docstring to explain what the return values are? Finally, is there any reason to return `None` for the flex_cols but `{}` for the record? Seems like we should do one or the other but not both.

File: dataactvalidator/filestreaming/csvAbstractReader.py [hunk 3]
```diff
@@ -32,6 +32,7 @@ def open_file(self, region, bucket, filename, csv_schema, bucket_name, error_fil
         self.unprocessed = ''
         self.extra_line = False
         self.lines = []
+        self.flex_dictionary = {}
         self.header_dictionary = {}
```
Comment (cmc333333) [inline dataactvalidator/filestreaming/csvAbstractReader.py]: While we're here, does it make sense to replace these with lists? It looks like we're only indexing them by incrementing integer values.

File: dataactvalidator/validation_handlers/validationManager.py [hunk 4]
```diff
@@ -145,9 +146,14 @@ def readRecord(self,reader,writer,file_type,interfaces,row_number,job_id,fields,
...
+            record = FieldCleaner.cleanRow(next_record, self.long_to_short_dict, fields)
             record["row_number"] = row_number
+            if not flex_cols:
+                flex_cols = None
+            else:
+                flex_cols["row_number"] = row_number
+
             if reader.is_finished and len(record) < 2:
                 # This is the last line and is empty, don't record an error
                 return {}, True, True, True, False  # Don't count this row
```
Comment (cmc333333) [inline dataactvalidator/validation_handlers/validationManager.py]: Does this need a sixth element?

File: dataactvalidator/validation_handlers/validationManager.py [hunk 5]
```diff
@@ -240,6 +247,30 @@ def writeErrors(self, failures, interfaces, job_id, short_colnames, writer, warn
...
+            Boolean indicating whether to skip current row
+        """
+        sess = GlobalDB.db().session
+        try:
+            flex_cols["job_id"] = job_id
+            flex_cols["submission_id"] = submission_id
+
+            sess.add(FlexField(**flex_cols))
+            sess.commit()
+        except ResponseException as e:
```
Comment (cmc333333) [inline dataactvalidator/validation_handlers/validationManager.py]: I don't think the above would raise a ResponseException -- is there some sqlalchemy exception you want to catch instead?

File: dataactvalidator/validation_handlers/validator.py [hunk 6]
```diff
@@ -200,15 +197,27 @@ def validateFileBySql(cls, submissionId, fileType, short_to_long_dict):
             if failures.rowcount:
                 # Create column list (exclude row_number)
                 cols = failures.keys()
+
+                # Create flex column list
+                flex_dict = {}
+                flex_results = sess.execute("SELECT * FROM flex_field WHERE submission_id="+str(submissionId))
```
Comment (cmc333333) [inline dataactvalidator/validation_handlers/validator.py]: Would you mind using the ORM here? It'd make it much clearer to understand what fields are being accessed.

File: dataactvalidator/validation_handlers/validationManager.py (hunk 4 above)
Comment (cmc333333) [inline dataactvalidator/validation_handlers/validationManager.py]: Sure thing! The other two `return`s in this function are returning a six-element tuple, but this one is only returning a five-element tuple. I think you just missed one `return` when you added that sixth element to the other two.

**Comment by cmc333333** [comment]
Looks good! @michaeldhess has agreed to a follow up PR to add tests.

**Pull request merged**